packages/core/src/controllers/publisher.ts

```diff
@@ -57,52 +61,67 @@ export class Publisher extends IPublisher {
        */
-      while (result === undefined) {
-        // Terminate the publishing attempts if publisTimeout has been exceeded
-        if (Date.now() - startPublish > this.publishTimeout) {
-          throw new Error(failedPublishMessage);
-        }
-
-        this.logger.trace({ id, attempts }, `publisher.publish - attempt ${attempts}`);
-        const publish = await createExpiringPromise(
-          this.rpcPublish(topic, message, ttl, relay, prompt, tag, id, opts?.attestation).catch(
-            (e) => this.logger.warn(e),
-          ),
-          this.publishTimeout,
-          failedPublishMessage,
+      const publishPromise = new Promise(async (resolve) => {
+        const onPublish = ({ id }: { id: string }) => {
+          if (params.opts.id === id) {
+            this.removeRequestFromQueue(id);
+            this.relayer.events.removeListener(RELAYER_EVENTS.publish, onPublish);
+            resolve(params);
+          }
+        };
+        this.relayer.events.on(RELAYER_EVENTS.publish, onPublish);
+        const initialPublish = createExpiringPromise(
+          new Promise((resolve, reject) => {
+            this.rpcPublish({
+              topic,
+              message,
+              ttl,
+              prompt,
+              tag,
+              id,
+              attestation: opts?.attestation,
+            })
+              .catch((e) => {
+                this.logger.warn(e, e?.message);
+                reject(e);
+              })
+              .then(resolve);
+          }),
+          this.initialPublishTimeout,
+          `Failed initial publish, retrying.... id:${id} tag:${tag}`,
         );
-        result = await publish;
-        attempts++;
-
-        if (!result) {
-          // small delay before retrying so we can limit retry to max 1 time per second
-          // if network is down `rpcPublish` will throw immediately
-          await new Promise((resolve) => setTimeout(resolve, this.failedPublishTimeout));
-        }
-      }
-      this.relayer.events.emit(RELAYER_EVENTS.publish, params);
-      this.logger.debug(`Successfully Published Payload`);
+        await initialPublish
+          .then((result) => {
+            this.events.removeListener(RELAYER_EVENTS.publish, onPublish);
+            resolve(result);
+          })
+          .catch((e) => {
+            this.queue.set(id, { ...params, attempt: 1 });
+            this.logger.warn(e, e?.message);
+          });
+      });
       this.logger.trace({
         type: "method",
         method: "publish",
         params: { id, topic, message, opts },
       });
+
+      await createExpiringPromise(publishPromise, this.publishTimeout, failedPublishMessage);
     } catch (e) {
       this.logger.debug(`Failed to Publish Payload`);
       this.logger.error(e as any);
       if (opts?.internal?.throwOnFailedPublish) {
         throw e;
       }
-      this.queue.set(id, params);
+    } finally {
+      this.queue.delete(id);
```

Comment:
So previously we were setting the request in the queue so that it's retired but now we're deleting it from the queue because the retry logic is part of the `initialPublish` right ?

Reply:
correct. Now the request is set into the queue only after `initialPublish` fails